### PR TITLE
feat: Move to passport for authentication

### LINKED ISSUE #1120: Changing towards passport.js for authentication
Lots of different OAuth providers are being requested, e.g. in #755, #962, or some other issues.

I'm in the process of working for #962 and was wondering about moving towards passport.js, which would make for a much easier provider addition procedure.

### PR DESCRIPTION
**Pull request by tommoor**
Based on #1929
Supersedes https://github.com/outline/outline/pull/1183
Closes #1120

### CODE REVIEW

**Comment by ulentini** [comment]
Hi, this is a great feature, we really needed it. 
With very little effort and using a few ENV variable, self-hosted "single team" could be implemented (ie: all configured signin method would be always available and all users would join the same single team, like discussed on #862).
I'm already testing it locally and it works pretty fine, I think I'm gonna test it on a few of our outline installations.
Would you be interested in a PR about this?

**Comment by tommoor** [comment]
> "single team" could be implemented (ie: all configured sign-in method would be always available

Yes, ideally this would be the default behavior to be honest, but such a change will have to wait until this is all landed.

> on a few of our outline installations.

You have multiple installations? 🤔

**Comment by ulentini** [comment]
> > "single team" could be implemented (ie: all configured sign-in method would be always available
> 
> Yes, ideally this would be the default behavior to be honest, but such a change will have to wait until this is all landed.
> 
> > on a few of our outline installations.
> 
> You have multiple installations? 🤔
> You have multiple installations? 🤔
Yes 😄

File: server/commands/accountProvisioner.js [hunk 1]
```diff
@@ -0,0 +1,105 @@
...
+      subdomain: teamParams.subdomain,
+      avatarUrl: teamParams.avatarUrl,
+      authenticationProvider,
+    });
+  } catch (err) {
+    throw new AuthenticationError(err.message);
+  }
+  invariant(team, "Team must exist");
+
+  const authP = team.authenticationProviders[0];
```
Comment (thenanyu) [inline server/commands/accountProvisioner.js:57]: feels unsafe — this is an explicit guarantee but it violates demeter — better to use `_.find(match the provider id)`

File: server/commands/accountProvisioner.js [hunk 2]
```diff
@@ -0,0 +1,105 @@
...
+          email: userParams.email,
+          teamId: team.id,
+        },
+      });
+
+      if (exists) {
+        throw new EmailAuthenticationRequiredError(
+          "Email authentication required",
+          team.url
+        );
```
Comment (thenanyu) [inline server/commands/accountProvisioner.js:97]: this is slightly suspect but at worst it's doing pretty much nothing, maybe note it in a comment or something

**Comment by jakoblorz** [comment]
Great work on this!

**Comment by jdaviescoates** [comment]
This is awesome! @tommoor will you be releasing a new version that includes this in the near future? Many thanks for your work on this!

**Comment by tommoor** [comment]
New releases are cut once/month, generally around the ~15th depending on what's in flight.

**Comment by VityaSchel** [comment]
image

guys.... it's been more than a year... can someone tell me is this feature in beta or like how to enable it? me and some other people are still subscribed and waiting for this feature, what is the current status? was it implemented? is there something we have to change in docker.env?

**Comment by VityaSchel** [comment]
image

wtf

**Pull request merged**